### PR TITLE
GH-43454: [C++][Python] Add Opaque canonical extension type

### PR DESCRIPTION
**Pull request by lidavidm**
### Rationale for this change

Add the newly ratified extension type.

### What changes are included in this PR?

The C++/Python implementation only.

### Are these changes tested?

Yes

### Are there any user-facing changes?

No.
* GitHub Issue: #43454

### CODE REVIEW

**Review by jorisvandenbossche — COMMENTED** [review]
Copying my review comments from https://github.com/apache/arrow/pull/41823#pullrequestreview-2196033541

For the Python bindings, looks good as well, just a few minor comments:

- We probably want to add `OpaqueScalar` to `pyarrow/__init__.py` as well? (although I see we forgot this for the tensor type, which I assume you have been imitating)
- Can you add the three new classes to `test_extension_type_constructor_errors` in `test_misc.py`?
- Can you list the new type in the python API docs (docs/source/python/api/datatypes.rst, docs/source/python/api/arrays.rst
)? See eg [#39652](https://github.com/apache/arrow/pull/39652/files) for a previous PR adding a new type where this is added to the docs.

File: python/pyarrow/tests/test_extension_type.py [hunk 1]
```diff
@@ -1661,3 +1661,49 @@ def test_legacy_int_type():
...
+    assert opaque_type != pa.opaque(storage_type, "type2", "vendor")
+    assert opaque_type != pa.opaque(storage_type, "type", "vendor2")
+    assert opaque_type != pa.opaque(pa.decimal128(12, 3), "type", "vendor")
+
+    # Pickle roundtrip
+    result = pickle_module.loads(pickle_module.dumps(opaque_type))
+    assert result == opaque_type
+
+    # IPC roundtrip
+    tensor_arr_class = opaque_type.__arrow_ext_class__()
```
Comment (joellubi) [inline python/pyarrow/tests/test_extension_type.py:1691]: `opaque_arr_class`?

**Comment by lidavidm** [comment]
I believe I've fixed everything here, thanks for the reviews!

**Review by jorisvandenbossche — APPROVED** [review]
Haven't looked in detail at the C++ side again, but the Python changes look good!

**Comment by lidavidm** [comment]
Any further comments here?

**Comment by joellubi** [comment]
> Any further comments here?

None for me. I based some of my work on Bool8 on the implementation here, so ideally we could get another C++ reviewer to approve as a third party.

File: cpp/src/arrow/extension/opaque_test.cc [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+  ASSERT_NE(*type2, *type3);
+  ASSERT_NE(*type2, *type4);
+
+  ASSERT_NE(*type3, *type);
+  ASSERT_NE(*type3, *type2);
+  ASSERT_NE(*type3, *type4);
+
+  ASSERT_NE(*type4, *type);
+  ASSERT_NE(*type4, *type2);
+  ASSERT_NE(*type4, *type3);
```
Comment (pitrou) [inline cpp/src/arrow/extension/opaque_test.cc:87]: Perhaps also compare with an extension type that is not Opaque?

**Review by pitrou — COMMENTED** [review]
LGTM, a couple minor comments

File: cpp/src/arrow/extension/opaque_test.cc (hunk 2 above)
Comment (lidavidm) [reply cpp/src/arrow/extension/opaque_test.cc:93]: Updated.

**Comment by lidavidm** [comment]
I'm going to merge this in a bit given the radio silence

**Comment by rok** [comment]
Apologies for the late question. Is opaque type intentionally not registered?
https://github.com/apache/arrow/blob/f078942ce2df68de8f48c3b4233132133601ca53/cpp/src/arrow/extension_type.cc#L145-L157
If it's an oversight I can add registration in https://github.com/apache/arrow/pull/37298.

**Comment by lidavidm** [comment]
It was an oversight, thanks for spotting it

**Comment by jorisvandenbossche** [comment]
> If it's an oversight I can add registration in #37298.

Let's do that as a separate PR (it should also require to update the tests slightly, and then it doesn't depend on the timeline of merging the other PR)

**Comment by jorisvandenbossche** [comment]
Opened https://github.com/apache/arrow/issues/43787 to keep track of it.

**Pull request merged**